run_mount.go

```diff
@@ -41,6 +51,50 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
 		anonVolumes []string
 	)
 	mounted := make(map[string]struct{})
+	var imageVolumes map[string]struct{}
+	var tempDir string
+
+	if ensuredImage != nil {
+		imageVolumes = ensuredImage.ImageConfig.Volumes
+
+		if err := ensuredImage.Image.Unpack(ctx, clicontext.String("snapshotter")); err != nil {
+			return nil, nil, errors.Wrap(err, "error unpacking image")
+		}
+
+		diffIDs, err := ensuredImage.Image.RootFS(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		chainID := identity.ChainID(diffIDs).String()
+
+		s := client.SnapshotService(clicontext.String("snapshotter"))
+		tempDir, err = ioutil.TempDir("", "initialC")
+		if err != nil {
+			return nil, nil, err
+		}
+		// We use Remove here instead of RemoveAll.
+		// The RemoveAll will delete the temp dir and all children it contains.
+		// When the Unmount fails, RemoveAll will incorrectly delete data from the mounted dir
+		defer os.Remove(tempDir)
+
+		var mounts []mount.Mount
+		mounts, err = s.View(ctx, tempDir, chainID)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// We should do defer first, if not we will not do Unmount when only a part of Mounts are failed.
+		defer func() {
+			err = mount.UnmountAll(tempDir, 0)
+		}()
+
+		if err := mount.All(mounts, tempDir); err != nil {
+			if err := s.Remove(ctx, tempDir); err != nil && !errdefs.IsNotFound(err) {
```

Comment:
need to umount first and then remove

Reply:
https://github.com/containerd/nerdctl/pull/216